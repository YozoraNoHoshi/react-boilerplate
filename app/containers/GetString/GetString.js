import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Flex from '../../components/Flex';

class GetStrings extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      errorMessage: '',
    };
  }

  // We treat the database as the source of truth, hence why we want to ensure the redux store is up to date with the current database.
  componentDidMount() {
    try {
      this.props.getStrings();
    } catch (error) {
      this.setState({ errorMessage: error.message });
    }
  }

  showStrings = strings => {
    const stringItems = strings.map(s => <li key={s}>{s}</li>);
    return <ul>{stringItems}</ul>;
  };

  render() {
    return (
      <Flex
        sMargin="10px"
        sPadding="10px"
        style={{
          boxShadow: '2px 2px 5px black',
        }}
      >
        <Flex large>Here are all the strings we have in our database!</Flex>
        {this.showStrings(this.props.strings)}
        {this.state.errorMessage && <div>{this.state.errorMessage}</div>}
      </Flex>
    );
  }
}

GetStrings.defaultProps = { getStrings: () => {}, strings: [] };

GetStrings.propTypes = { strings: PropTypes.array, getStrings: PropTypes.func };

export default GetStrings;
