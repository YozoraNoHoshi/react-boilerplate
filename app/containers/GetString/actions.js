import axios from 'axios';
import { GET_STRING, GET_STRING_TO_STORE } from '../../reducers/constants';

const BASE_URL = 'http://localhost:3000/api';

export function getStringAction() {
  return { type: GET_STRING };
}

export function getStringsToStore(payload) {
  return { type: GET_STRING_TO_STORE, payload };
}

export async function getStrings() {
  try {
    const result = await axios.get(BASE_URL);
    return result.data.strings;
  } catch (error) {
    throw new Error(error.response.data.message);
  }
}
