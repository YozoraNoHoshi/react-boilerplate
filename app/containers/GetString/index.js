import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import injectSaga from 'utils/injectSaga';
import { RESTART_ON_REMOUNT } from 'utils/constants';
import saga from './sagas';
import { getStringAction } from './actions';
import { selectStrings } from '../../selectors';

class GetStrings extends PureComponent {
  componentDidMount() {
    this.props.getStrings();
  }

  showStrings = strings => {
    const stringItems = strings.map(s => <li key={s}>{s}</li>);
    return <ul>{stringItems}</ul>;
  };

  render() {
    return (
      <div className="GetStrings">
        <div>I hate you</div>
        {this.showStrings(this.props.strings)}
      </div>
    );
  }
}

GetStrings.propTypes = { strings: PropTypes.array, getStrings: PropTypes.func };

function mapStateToProps(state) {
  return { strings: selectStrings(state) };
}

const withSaga = injectSaga({
  key: 'getStrings',
  saga,
  mode: RESTART_ON_REMOUNT,
});

const withConnect = connect(
  mapStateToProps,
  { getStrings: getStringAction },
);

export default compose(
  withConnect,
  withSaga,
)(GetStrings);
