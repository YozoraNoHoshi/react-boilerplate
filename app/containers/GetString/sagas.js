import { call, put, takeLatest } from 'redux-saga/effects';
import { getStrings } from './actions';
import { GET_STRING, GET_STRING_TO_STORE } from '../../reducers/constants';

// worker Saga: will be fired on USER_FETCH_REQUESTED actions
function* getStringsFromDB() {
  const string = yield call(getStrings);
  yield put({ type: GET_STRING_TO_STORE, payload: string.strings });
}

function* getStringsSaga() {
  yield takeLatest(GET_STRING, getStringsFromDB);
}

export default getStringsSaga;
