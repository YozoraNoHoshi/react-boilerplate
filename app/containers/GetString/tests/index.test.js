import React from 'react';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';
import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import GetString from '../GetString';
import db from '../../../../server/db';
import ConnectedGetString from '../index';
import configureStore from '../../../configureStore';

configure({ adapter: new Adapter() });

beforeAll(async () => {
  await db.query(`DELETE FROM strings`);
  await db.query(`INSERT INTO strings (string) VALUES ($1), ($2), ($3)`, [
    'bleh',
    'aaaaa',
    'bbbbbb',
  ]);
});

describe('<GetString /> Component', () => {
  it('should render and match the snapshot', () => {
    const wrapper = shallow(<GetString strings={['aaaa', 'bbbbb', 'ccc']} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

// Since the local backend doesn't seem to start up during testing environment, this test is disabled because it will not be an accurate snapshot
// The snapshot is of a failed request

describe('<GetString /> Connected to Redux', () => {
  it('should render and match the snapshot', () => {
    const store = configureStore({}, browserHistory);
    const wrapper = mount(
      <Provider store={store}>
        <ConnectedGetString />
      </Provider>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

// Do not need to test the componentDidMount because that is handled in sagas.test.js
// All we need to test here is that the component renders and renders its prop 'strings' properly.

afterAll(async () => {
  await db.query(`DELETE FROM strings`);
  await db.end();
});
