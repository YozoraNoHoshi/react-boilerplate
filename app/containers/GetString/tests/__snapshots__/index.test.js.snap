// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GetString /> Component should render and match the snapshot 1`] = `
<Flex
  sMargin="10px"
  sPadding="10px"
  style={
    Object {
      "boxShadow": "2px 2px 5px black",
    }
  }
>
  <Flex
    large={true}
  >
    Here are all the strings we have in our database!
  </Flex>
  <ul>
    <li
      key="aaaa"
    >
      aaaa
    </li>
    <li
      key="bbbbb"
    >
      bbbbb
    </li>
    <li
      key="ccc"
    >
      ccc
    </li>
  </ul>
</Flex>
`;

exports[`<GetString /> Connected to Redux should render and match the snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "injectedReducers": Object {},
      "injectedSagas": Object {
        "getStrings": Object {
          "mode": "@@saga-injector/restart-on-remount",
          "saga": [Function],
          "task": Object {
            "@@redux-saga/TASK": true,
            "cancel": [Function],
            "cont": [Function],
            "context": Object {},
            "end": [Function],
            "error": [Function],
            "id": 1,
            "isAborted": [Function],
            "isCancelled": [Function],
            "isRoot": true,
            "isRunning": [Function],
            "joiners": Array [],
            "meta": Object {
              "location": undefined,
              "name": "getStringsSaga",
            },
            "queue": Object {
              "abort": [Function],
              "addTask": [Function],
              "cancelAll": [Function],
              "getTasks": [Function],
            },
            "result": [Function],
            "setContext": [Function],
            "toPromise": [Function],
          },
        },
      },
      "replaceReducer": [Function],
      "runSaga": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ConnectFunction>
    <withSaga(GetStrings)
      getStrings={[Function]}
      strings={Array []}
    >
      <GetStrings
        getStrings={[Function]}
        strings={Array []}
      >
        <Flex
          sMargin="10px"
          sPadding="10px"
          style={
            Object {
              "boxShadow": "2px 2px 5px black",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Flex-sc-692hpc-0",
                  "isStatic": false,
                  "lastClassName": "fOnmvU",
                  "rules": Array [
                    "display:flex;flex-direction:",
                    [Function],
                    ";justify-content:",
                    [Function],
                    ";align-items:",
                    [Function],
                    ";",
                    [Function],
                    " font-size:",
                    [Function],
                    ";",
                    [Function],
                    " ",
                    [Function],
                  ],
                },
                "displayName": "Flex",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Flex-sc-692hpc-0",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            sMargin="10px"
            sPadding="10px"
            style={
              Object {
                "boxShadow": "2px 2px 5px black",
              }
            }
          >
            <div
              className="Flex-sc-692hpc-0 kmIEHT"
              style={
                Object {
                  "boxShadow": "2px 2px 5px black",
                }
              }
            >
              <Flex
                large={true}
              >
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "Flex-sc-692hpc-0",
                        "isStatic": false,
                        "lastClassName": "fOnmvU",
                        "rules": Array [
                          "display:flex;flex-direction:",
                          [Function],
                          ";justify-content:",
                          [Function],
                          ";align-items:",
                          [Function],
                          ";",
                          [Function],
                          " font-size:",
                          [Function],
                          ";",
                          [Function],
                          " ",
                          [Function],
                        ],
                      },
                      "displayName": "Flex",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "Flex-sc-692hpc-0",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  large={true}
                >
                  <div
                    className="Flex-sc-692hpc-0 fOnmvU"
                  >
                    Here are all the strings we have in our database!
                  </div>
                </StyledComponent>
              </Flex>
              <ul />
            </div>
          </StyledComponent>
        </Flex>
      </GetStrings>
    </withSaga(GetStrings)>
  </ConnectFunction>
</Provider>
`;
